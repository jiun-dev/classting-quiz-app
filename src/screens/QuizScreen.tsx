import React, {useEffect, useState} from 'react';
import {View, Text, Alert, StyleSheet, TouchableOpacity} from 'react-native';
import {getQuizList} from '../api/quizApi';
import BasicButton from '../components/BasicButton';
import {SafeAreaView} from 'react-native-safe-area-context';

import StatusBar from '../components/StatusBar';
import BaseLoading from '../components/Loading';
import {StackScreenProps} from '@react-navigation/stack';
import {colors, fontSize, radius, spacing} from '../constants/\btheme';

import AsyncStorage from '@react-native-async-storage/async-storage';
import {decodeHTMLEntities} from '../utils/decodeHtml';

export type QuizScreenProps = StackScreenProps<RootNavigationType, 'Quiz'>;

function QuizScreen({navigation}: QuizScreenProps) {
  const [currentQuestionIndex, setCurrentQuestionIndex] = useState(0);
  const [quizQuestions, setQuizQuestions] = useState<Question[]>([]);
  const [wrongAnswers, setWrongAnswers] = useState<Question[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [correctCount, setCorrectCount] = useState(0);
  const [selectedAnswer, setSelectedAnswer] = useState<string | null>(null);
  const [userAnswers, setUserAnswers] = useState<string[]>([]);
  const [isAnswerChecked, setIsAnswerChecked] = useState(false);
  const [timer, setTimer] = useState(0);

  useEffect(() => {
    getchQuizData();
    setupTimer();
  }, []);

  const getchQuizData = async () => {
    try {
      const data = await getQuizList();
      setQuizQuestions(data);
    } catch (error) {
      console.error('ERROR FETCHING QUIZ DATA : ', error);
      Alert.alert(
        'ERROR',
        '퀴즈 정보를 가져오는데 실패했습니다.\n잠시 후 시도해주세요',
      );
    }
    setIsLoading(false);
  };

  const setupTimer = () => {
    const interval = setInterval(() => {
      setTimer(prevTime => prevTime + 1);
    }, 1000);
    return () => clearInterval(interval);
  };

  const handleAnswer = () => {
    if (!selectedAnswer) {
      Alert.alert('알림', '정답을 선택해주세요.');
      return;
    }

    setIsAnswerChecked(true);
    const isAnswerCorrect = selectedAnswer === currentQuestion.correct_answer;
    if (isAnswerCorrect) {
      setCorrectCount(correctCount + 1);
    } else {
      setWrongAnswers([...wrongAnswers, currentQuestion]);
    }
    const updatedAnswers = [...userAnswers];
    updatedAnswers[currentQuestionIndex] = selectedAnswer;
    setUserAnswers(updatedAnswers);
  };

  const handleNextQuestion = () => {
    if (currentQuestionIndex < quizQuestions.length - 1) {
      setCurrentQuestionIndex(currentQuestionIndex + 1);
    } else {
      finishQuiz();
    }
    setSelectedAnswer(null);
    setIsAnswerChecked(false);
  };

  const finishQuiz = async () => {
    const wrongResults = quizQuestions
      .map((question, index) => ({
        ...question,
        userAnswer: userAnswers[index],
        isCorrect: userAnswers[index] === question.correct_answer,
        questionId: index + 1,
      }))
      .filter(question => !question.isCorrect);

    const quizResult = {
      date: new Date().toISOString(),
      wrongAnswers: wrongResults,
    };

    try {
      const existingResults = await AsyncStorage.getItem('wrongQuizResults');
      const updatedResults = existingResults
        ? [...JSON.parse(existingResults), quizResult]
        : [quizResult];
      await AsyncStorage.setItem(
        'wrongQuizResults',
        JSON.stringify(updatedResults),
      );
      navigation.replace('Result', {
        correctCount,
        totalQuestions: quizQuestions.length,
        elapsedTime: timer,
      });
    } catch (error) {
      console.error('저장 실패', error);
    }
  };

  const currentQuestion = quizQuestions[currentQuestionIndex];

  if (isLoading || !quizQuestions.length) {
    return <BaseLoading />;
  }

  const renderOptionsView = () => {
    const options = currentQuestion.incorrect_answers.concat(
      currentQuestion.correct_answer,
    );
    return options.map((option, index) => (
      <TouchableOpacity
        key={index}
        style={styles.option}
        onPress={() => setSelectedAnswer(option)}>
        <View
          style={
            selectedAnswer === option ? styles.radioSelected : styles.radio
          }
        />
        <Text style={styles.optionText}>{option}</Text>
      </TouchableOpacity>
    ));
  };

  const renderResultView = () => {
    return (
      <Text
        style={
          selectedAnswer === currentQuestion.correct_answer
            ? styles.correctAnswer
            : styles.wrongAnswer
        }>
        {selectedAnswer === currentQuestion.correct_answer
          ? '정답입니다!'
          : '틀렸습니다!'}
      </Text>
    );
  };

  return (
    <SafeAreaView style={styles.container}>
      <StatusBar
        elapsedTime={timer}
        currentQuestionIndex={currentQuestionIndex}
        totalQuestions={quizQuestions.length}
      />
      <Text style={styles.question}>
        {decodeHTMLEntities(currentQuestion.question)}
      </Text>
      {isAnswerChecked ? (
        <>
          {renderResultView()}
          <View style={styles.buttonContainer}>
            <BasicButton title="다음" onPress={() => handleNextQuestion()} />
          </View>
        </>
      ) : (
        <>
          {renderOptionsView()}
          <BasicButton title="확인" onPress={() => handleAnswer()} />
        </>
      )}
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: spacing.medium,
  },
  correctAnswer: {
    color: colors.main,
    fontSize: fontSize.large,
  },
  wrongAnswer: {
    color: colors.error,
    fontSize: fontSize.large,
  },
  question: {
    fontSize: fontSize.xxl,
    fontWeight: 'bold',
    marginVertical: spacing.medium,
  },
  option: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: spacing.medium,
  },
  radio: {
    height: spacing.medium,
    width: spacing.medium,
    borderRadius: radius.medium,
    borderWidth: 2,
    borderColor: colors.black,
    marginRight: spacing.medium,
  },
  radioSelected: {
    height: spacing.medium,
    width: spacing.medium,
    borderRadius: radius.medium,
    borderWidth: 2,
    borderColor: colors.main,
    marginRight: spacing.medium,
  },
  optionText: {},
  buttonContainer: {
    marginVertical: spacing.small,
  },
});

export default QuizScreen;
