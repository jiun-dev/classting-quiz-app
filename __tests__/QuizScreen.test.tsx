import {render} from '@testing-library/react-native';
import QuizScreen from '../src/screens/QuizScreen';
import {StackNavigationProp} from '@react-navigation/stack';
import {RouteProp} from '@react-navigation/native';

describe('QuizScreen 컴포넌트 테스트', () => {
  const mockNavigate = jest.fn();
  const props: {
    navigation: StackNavigationProp<RootNavigationType, 'Quiz'>;
    route: RouteProp<RootNavigationType, 'Quiz'>;
  } = {
    navigation: {
      navigate: mockNavigate,
      goBack: jest.fn(),
      addListener: jest.fn(),
      removeListener: jest.fn(),
      setOptions: jest.fn(),
    } as any,
    route: {
      key: 'mockKey',
      name: 'Quiz',
      params: {},
    },
  };

  test('컴포넌트 렌더링 테스트', () => {
    const rendered = render(<QuizScreen {...props} />);

    // Add assertions to verify if specific elements are rendered
    // For example, you can check for the presence of certain text or buttons
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});
