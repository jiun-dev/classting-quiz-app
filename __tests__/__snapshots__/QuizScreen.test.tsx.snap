// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizScreen 컴포넌트 테스트 컴포넌트 렌더링 테스트 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    indeterminateAnimationDuration={1000}
    style={
      [
        {
          "backgroundColor": "transparent",
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <RNSVGSvgView
      bbHeight="30"
      bbWidth="30"
      collapsable={false}
      fill="none"
      focusable={false}
      height={30}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "transform": [
              {
                "rotate": "0deg",
              },
            ],
          },
          {
            "flex": 0,
            "height": 30,
            "width": 30,
          },
        ]
      }
      width={30}
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M15 0.5
          A14.5 14.5 0 1 1 6.477113841759136 3.269253581563264"
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={
            {
              "payload": 4283215696,
              "type": 0,
            }
          }
          strokeLinecap={0}
          strokeWidth="1"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
